src/metrics/metrics.ts

```diff
@@ -65,79 +66,81 @@ export function countGamesInDataset(datasetPath: string): number {
-  const moveHistory = chess.history();
-
-  // duplicate the base map
-  const pieceSquares = new Map<Square, UnambiguousPieceSymbol>(
-    basePieceSquares
-  );
+  const chess = new Chess();
+  const moveGenerator = chess.historyGenerator(game.moves);
 
   // create an object to track distance value for each piece
   const distanceMap: { [key: string]: number } = {};
-  for (const piece of pieceSquares.values()) {
-    distanceMap[piece] = 0;
-  }
+
+  let lastMove: PrettyMove | null = null;
 
   // Initialize variables to keep track of the maximum distance and the piece
   let maxDistance = -1;
   let maxDistancePiece: UnambiguousPieceSymbol;
 
-  // TODO: we'll need to update the labels we use in cjsmin to be unique to do things this way
-  for (const { from, to } of moveHistory) {
-    // Calculate the file (column) distance by subtracting ASCII values
-    const fileDist = Math.abs(from.charCodeAt(0) - to.charCodeAt(0));
-    // Calculate the rank (row) distance by subtracting numeric values
-    const rankDist = Math.abs(Number(from[1]) - Number(to[1]));
-    // The distance moved is the maximum of fileDist and rankDist
-    const distance = Math.max(fileDist, rankDist);
-    // Get the piece that moved from the pieceSquares map
-    const movedPiece = pieceSquares.get(from);
-
-    // we'll update the map as pieces move. To avoid additional operations we only update (no delete). Can change this if we use this elsewhere
-    pieceSquares.set(to, movedPiece);
-    distanceMap[movedPiece] += distance;
-
-    if (distanceMap[movedPiece] > maxDistance) {
-      maxDistance = distanceMap[movedPiece];
-      maxDistancePiece = movedPiece;
+  // evaluate each move, update the correct unambiguous piece's distance
+  for (let moveInfo of moveGenerator) {
+    const { move, board } = moveInfo;
+
+    lastMove = move;
+    let piece = lastMove.unambiguousSymbol;
+
+    if (!distanceMap[piece]) {
+      distanceMap[piece] = 0;
+    }
+
+    const fromMove = moveInfo.move.from
+    const toMove = moveInfo.move.to
+
+    let distance = 0;
+
+    // Check if the move is a castling move
+    if (moveInfo.move.flags === 'k' || moveInfo.move.flags === 'q') {
+      // move was done by black
+      if (moveInfo.move.color === 'b'){
+        // queenside castle
+        if (moveInfo.move.flags === 'q') {
+          distanceMap['k'] += 2
+          distanceMap['ra'] = (distanceMap['ra'] || 0) + 3
+        }
+        //kingside castle
+        else {
+          distanceMap['k'] += 2
+          distanceMap['rh'] = (distanceMap['rh'] || 0) + 2
+        }
+      }
+      // move was done by white
+      else {
+        // queenside castle
+        if (moveInfo.move.flags === 'q') {
+          distanceMap['K'] += 2
+          distanceMap['RA'] = (distanceMap['RA'] || 0) + 3
+        }
+        // kingside castle
+        else {
+          distanceMap['K'] += 2
+          distanceMap['RH'] = (distanceMap['RH'] || 0) + 2
+        }
+      }
```

Comment:
Not strictly necessary but this if is hard to follow. I'd propose this to make it more readable?
```suggestion
let movingKing =  'k'
let movingRook, rookDistance;

if (moveInfo.move.flags === 'k') {
rookDistance = 2
movingRook = 'ra'
} else {
rookDistance = 3
movingRook = 'rh'
}
if (moveInfo.move.color === 'w') {
movingRook = movingRook.toUpperCase()
movingKing = movingKing.toUpperCase()
}

distanceMap[movingKing] += 2
distanceMap[movingRook] += rookDistance
```
Something along these lines

Reply:
hacked this together, this is highely opinionated so dw too much but having less lines and no nesting makes it easier to follow imo

Reply:
yeah fair enough. smart solution for sure but idk how much i care to do stuff like this, feel free if you have high preference. i do think urs is more elegant